lib/surface/compiler/parser.ex

```diff
@@ -117,71 +124,76 @@ defmodule Surface.Compiler.Parser do
+      # pop the current buffer and use it as children for the sub-block node
+      [buffer | buffers] = buffers
+      node = {name, translated_attrs, Enum.reverse(buffer), to_meta(meta)}
+      buffers = push_node_to_current_buffer(node, buffers)
+      state = %{state | tags: tags}
+
+      {:ok, buffers, state}
+    end
   end
 
   # If there's no previous sub-block defined. Create a :default sub-block,
   # move the buffer there and close it.
   defp close_sub_block(token, buffers, %{tags: [{:tag_open, name, attrs, meta} | tags]} = state) do
-    validate_sub_block!(token, name)
-
-    # pop the current buffer and use it as children for the :default sub-block node
-    [buffer | buffers] = buffers
-    node = {:default, [], Enum.reverse(buffer), %{}}
+    with :ok <- sub_block_valid?(token, name) do
+      # pop the current buffer and use it as children for the :default sub-block node
+      [buffer | buffers] = buffers
+      node = {:default, [], Enum.reverse(buffer), %{}}
 
-    # create a new buffer for the parent node to replace the one that was popped
-    buffers = [[] | buffers]
-    buffers = push_node_to_current_buffer(node, buffers)
+      # create a new buffer for the parent node to replace the one that was popped
+      buffers = [[] | buffers]
+      buffers = push_node_to_current_buffer(node, buffers)
 
-    # push back the parent token to state
-    meta = Map.put(meta, :has_sub_blocks?, true)
-    state = %{state | tags: [{:tag_open, name, attrs, meta} | tags]}
+      # push back the parent token to state
+      meta = Map.put(meta, :has_sub_blocks?, true)
+      state = %{state | tags: [{:tag_open, name, attrs, meta} | tags]}
 
-    {buffers, state}
+      {:ok, buffers, state}
+    end
   end
 
   defp close_sub_block({:tag_open, name, _attrs, meta}, _buffers, _state) do
-    %{line: line, column: column} = meta
     valid_parents_str = message_for_invalid_sub_block_parent(name)
 
     message = "no valid parent node defined for <#{name}>. #{valid_parents_str}"
-    raise %ParseError{line: line, column: column, message: message}
+    {:error, %ParseError{file: meta.file, line: meta.line, column: meta.column, message: message}}
   end
 
-  defp validate_sub_block!({:tag_open, name, _attrs, meta}, parent_name) do
-    %{line: line, column: column} = meta
+  defp sub_block_valid?({:tag_open, name, _attrs, meta}, parent_name) do
     valid_parents_str = message_for_invalid_sub_block_parent(name)
 
-    if parent_name not in @sub_blocks_valid_parents[name] do
-      message = "cannot use <#{name}> inside <#{parent_name}>. #{valid_parents_str}"
-      raise %ParseError{line: line, column: column, message: message}
+    if parent_name in @sub_blocks_valid_parents[name] do
+      :ok
+    else
+      {:error,
+       %ParseError{
+         file: meta.file,
+         line: meta.line,
+         column: meta.column,
+         message: "cannot use <#{name}> inside <#{parent_name}>. #{valid_parents_str}"
+       }}
```

Comment:
I think this is probably the only thing I'd avoid from this PR. Propagating errors from inner function calls in a recursive algorithm is probably one of the best usages of exceptions, IMO. In our case, since any exception always halts the parser, we don't have to worry about performance so I think the whole logic is much easier to follow without all those `with` that had to be introduced to handle the returned error.

Reply:
I've switched it over to just raising exceptions and agree that it simplifies the logic considerably - i think my issue wasn't the choice between the two, and it wasn't clear to my why we raised in one place and return ok/error tuples in another. lmk what you think about the new commits

Reply:
> it wasn't clear to my why we raised in one place and return ok/error tuples in another

That's my fault. I should've mentioned the reason in the last message. The reason is that in some cases, depending on what we're parsing, we may want to have some kind of "checkpoint" where we can identify the context of what is being parsed. For instance, the `handle_interpolation` function is reused in 3 different places, each one in different contexts:

1. Classical interpolation in the body

https://github.com/surface-ui/surface/blob/175a5427e9e7a83a0493b3c4bfa4fb4d7f3e12ce/lib/surface/compiler/tokenizer.ex#L77-L87

2. Root property expression

https://github.com/surface-ui/surface/blob/175a5427e9e7a83a0493b3c4bfa4fb4d7f3e12ce/lib/surface/compiler/tokenizer.ex#L260-L269

3. Attribute value expression

https://github.com/surface-ui/surface/blob/175a5427e9e7a83a0493b3c4bfa4fb4d7f3e12ce/lib/surface/compiler/tokenizer.ex#L434-L444

By returning a tuple here, we explicitly indicate to the caller that this call, unlike `handle_text`, must return the expected "complete" token, if not, it's up to the caller to properly handle the error, for instance, to provide a more specific message based on the context of the call (we're not doing that yet but we will). Or to ignore the line/column where the parsing error happened and emit it pointing to the beginning of the token, if that makes more sense to the user. 

This kind of fine-grained error handling is one of the things it was really hard to achieve with `nimble_parcec`. With the custom tokenizer, we have a chance to define those checkpoints where we know the call must return a specific token. That also makes it easier to reason about the behaviour of the tokenizer during the recursion. Function calls returning tuples will not move the flow to other top-level handlers. They are complete. They either return the expected token or an error. They can't call, for instance, `handle_text` again.

So, here are some questions that can help to identify those special token handlers:

  * Must the call return a specific token? Or can it move the flow back to an upper-level call (restarting the recursion)?
  * Can it be reused by other handlers in different contexts?
  * Could an internal parsing error provided by an inner call provide sub-optimal information (message or position) to the user?

If the answer to any of those question is "yes", then there's a big change we should return a tuple.

Reply:
thanks, that explanation helps me understand your reasoning a lot better!